### PR TITLE
Enable ansi support by default on Windows 11+

### PR DESCRIPTION
**Pull request by plumstone**
# Enhancement proposal

Windows supports console ANSI output since build 10586 (End of 2015). Therefore I thought it would make sense to update the ANSI support detection in Spring Boot.

The only slight issue I noticed is that the JDK does not return the build number for Windows systems which in result cannot be checked. https://github.com/openjdk/jdk/blob/master/src/java.base/windows/native/libjava/java_props_md.c#L502

So with this change only the earliest versions of Windows 10 until 2016 (which are no longer supported anyway) would give out a wrong result whether they support ANSI.

Now I need your feedback. Is this acceptable? Should I add more test cases?

Thanks in advance. :)

### CODE REVIEW

**Comment by philwebb** [comment]
We've had a similar request in the past and found ANSI support in Windows quite hard to get working consistently. See https://github.com/spring-projects/spring-boot/pull/34835#issuecomment-3337323967 for some previous analysis.

If we do consider this again, I think it should be for Windows 11+ only.

**Comment by wilkinsona** [comment]
@plumstone would you like to update your proposed changes so that they're Windows 11+ only?

**Comment by snicoll** [comment]
@plumstone how is it going? No rush but wondering if that's still on your radar as we're approaching RC1.

**Comment by plumstone** [comment]
Implementation is done. I just tested the changes in a built bootJar. Works fine :)

image

File: core/spring-boot/src/main/java/org/springframework/boot/ansi/AnsiOutput.java [hunk 1]
```diff
@@ -171,13 +174,46 @@ private static boolean detectIfAnsiCapable() {
...
+	static boolean isWindows(String osName) {
+		return osName.toLowerCase(Locale.ENGLISH).contains("win");
+	}
+
+	private static final String CURRENT_BUILD = "CurrentBuild";
+	private static final Supplier<String> WINDOWS_BUILD_NUMBER = () -> {
+		try {
+			Process buildNumberRequest = Runtime.getRuntime().exec(
+					"reg query \"HKLM\\SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\" /v " + CURRENT_BUILD);
+			return new String(buildNumberRequest.getInputStream().readAllBytes());
```
Comment (wilkinsona) [inline core/spring-boot/src/main/java/org/springframework/boot/ansi/AnsiOutput.java]: I think shelling out like this might be a deal-breaker for me. Is there really no other way to detect Windows 11?

Comment (wilkinsona) [reply core/spring-boot/src/main/java/org/springframework/boot/ansi/AnsiOutput.java]: Looking at https://bugs.openjdk.org/browse/JDK-8274840, I think we could use the `os.name` system property looking for a value of `Windows 11` instead.

Comment (plumstone) [reply core/spring-boot/src/main/java/org/springframework/boot/ansi/AnsiOutput.java]: I can change it to os.name detection. But I don't know how long that will work reliably.

Comment (wilkinsona) [reply core/spring-boot/src/main/java/org/springframework/boot/ansi/AnsiOutput.java]: Yes please.

Given how [it's implemented](https://github.com/openjdk/jdk/commit/97ea9dd2f24f9f1fb9b9345a4202a825ee28e014) it should be no less reliable than what's proposed here.

Comment (plumstone) [reply core/spring-boot/src/main/java/org/springframework/boot/ansi/AnsiOutput.java]: Updated the branch. It was a bit trickier than I thought because the os name is quite different for windows server instances. But I added more tests which should cover all corner cases.

Comment (wilkinsona) [reply core/spring-boot/src/main/java/org/springframework/boot/ansi/AnsiOutput.java]: Thanks for the updates. Unfortunately, that's not quite what I had in mind. I think we should just look for the string `Windows 11` without any parsing. We could also look for `Windows Server 2025` which [was added in Java 24](https://github.com/openjdk/jdk/commit/34cddfbedd20d5804cab8044fbc402564e98eb9c). If the JDK team add more possible values to `os.name`, Boot can be updated to match at that time. I think that's preferable to the proposed parsing that adds quite a bit of complexity for no current benefit.

Comment (plumstone) [reply core/spring-boot/src/main/java/org/springframework/boot/ansi/AnsiOutput.java]: Ah okay now i get it :D

Will update the PR.

**Comment by plumstone** [comment]
Updated PR with the suggested changes
